tools/walletextension/wallet_extension.go

```diff
@@ -153,105 +134,80 @@ func (we *WalletExtension) handleHTTPEthJSON(resp http.ResponseWriter, req *http
 		return
 	}
 
-	// We unmarshall the JSON request.
-	var reqJSONMap map[string]interface{}
-	err = json.Unmarshal(body, &reqJSONMap)
-	if err != nil {
-		logAndSendErr(resp, fmt.Sprintf("could not unmarshall JSON-RPC request body to JSON: %s. "+
-			"If you're trying to generate a viewing key, visit %s", err, pathViewingKeys))
-		return
-	}
-	method := reqJSONMap[reqJSONKeyMethod]
-	fmt.Printf("Received %s request from wallet: %s\n", method, body)
-
-	reqJSONMap, err = we.ensureCallsHaveFromField(method, reqJSONMap)
+	rpcReq, err := parseRequest(body)
 	if err != nil {
 		logAndSendErr(resp, err.Error())
 		return
 	}
 
-	// We encrypt the request's params with the enclave's public key if it's a sensitive request.
-	maybeEncryptedBody, err := we.encryptParamsIfNeeded(body, method, reqJSONMap)
-	if err != nil {
-		logAndSendErr(resp, fmt.Sprintf("could not encrypt request parameters: %s", err))
-		return
-	}
-
-	// We forward the request on to the Obscuro node.
-	nodeResp, err := forwardMsgOverWebsocket(websocketProtocol+we.hostAddr, maybeEncryptedBody)
+	respStr := ""
+	err = we.hostClient.Call(&respStr, rpcReq.method, rpcReq.params...)
 	if err != nil {
-		logAndSendErr(resp, fmt.Sprintf("received error response when forwarding request to node at %s: %s", we.hostAddr, err))
+		logAndSendErr(resp, fmt.Sprintf("rpc request failed: %s", err))
 		return
 	}
 
-	// We unmarshall the JSON response.
-	var respJSONMap map[string]interface{}
-	err = json.Unmarshal(nodeResp, &respJSONMap)
-	if err != nil {
-		logAndSendErr(resp, fmt.Sprintf("could not unmarshall enclave response to JSON: %s", err))
-		return
-	}
+	respMap := make(map[string]interface{})
+	respMap[resJSONKeyID] = rpcReq.id
+	respMap[resJSONKeyRPCVer] = jsonrpc.Version
 
-	// We report any errors from the request.
-	if respJSONMap[respJSONKeyErr] != nil {
-		logAndSendErr(resp, respJSONMap[respJSONKeyErr].(map[string]interface{})[respJSONKeyMsg].(string))
-		return
-	}
+	// respStr could be a plain string value or could be a json encoded object
+	respMap[RespJSONKeyResult] = respStr
```

Comment:
@BedrockSquirrel I think it's clearer to have this as an else branch below, if the err isn't nil (so on line 163).